refactor fastscape extension

rlange2 opened the pull request:
Two [xarray-simlab](https://github.com/benbovy/xarray-simlab) processes that couple orographic precipitation and topography.

Issues:

1. it appears that `flowacc`, for now, has no effect on elevation
2. model crashes or runs very slow &#8594; possibly due to in imbalance of uplift and erosion 

ToDo:
- Solve issues
- Documentation
- change variable names
- supply default values for some variables
- Testing

Update:
- Issue#1: should be solved with the next FastScapeLib update (works with developer version, see [install fastscapelib-fortran from source](https://fastscape-lem.github.io/fastscapelib-fortran/#_install_the_python_package_from_source))
- Issue#2: will be resolved later

rlange2 committed:
refactor fastscape extension

rlange2 committed:
Introduce meaningful variables names

Input variables names have been changed to (hopefully) become more
intuitive. `attr` has been added to the list of imports to support
custom metadata with the `attrs` argument. Lastly, to maintain
consistency, the drainage process name has been changed.

rlange2 commented:
Apart from minor changes, the majority of input variable names has been changed into more meaningful values so that hopefully, the process itself becomes more intuitive to use.

For now, moist stability frequency `nm`, water vapour scale height `hw` and uplift sensitivity `cw` remain unchanged for the lack of alternative ideas. Also, this is how they are commonly referred to in the literature.

I'm happy this is taking shape and will soon fit the [fastscape ecosystem](https://github.com/fastscape-lem), @benbovy :)

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
-    tau_c = xs.variable(description="conversion time")
-    tau_f = xs.variable(description="fallout time")
+    lat = xs.variable(description="latitude")
+    precip_base = xs.variable(description="background precipitation")
```
```suggestion
    precip_base = xs.variable(description="background, nonorographic precipitation rate")
```

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
-    tau_f = xs.variable(description="fallout time")
+    lat = xs.variable(description="latitude")
+    precip_base = xs.variable(description="background precipitation")
+    wind_speed = xs.variable(description="windspeed")
```
```suggestion
    wind_speed = xs.variable(description="wind speed")
```

Add units.

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
+    lat = xs.variable(description="latitude")
+    precip_base = xs.variable(description="background precipitation")
+    wind_speed = xs.variable(description="windspeed")
+    wind_dir = xs.variable(description="wind direction")
```
```suggestion
    wind_dir = xs.variable(description="wind direction (azimuth)", attrs={"units": "degrees"})
```

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -35,18 +34,21 @@ class OrographicPrecipitation:
+            "tau_f" : self.fall_time,
+            "nm" : self.nm,
+            "hw" : self.hw,
+            "cw" : self.ref_density * self.lapse_rate_m / self.lapse_rate}
```
You could use the same parameter names in `compute_orographic_precip`.

benbovy replied:
Also, move the calculation of `cw` out of the dictionary. You have declared `cw` as a variable with intent='out' (which is good), but you don't provide any value for it:

```python
def _get_params(self):
    self.cw = self.ref_density * self.lapse_rate_m / self.lapse_rate

    return {
        "cw": self.cw
    }
````

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -1,3 +1,4 @@
+import attr
```
There is no need to import `attr` here.

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
-    rhosref = xs.variable(description="reference density")
+    lapse_rate = xs.variable(description="adiabatic lapse rate")
+    lapse_rate_m = xs.variable(description="moist adiabatic lapse rate")
+    ref_density = xs.variable(description="reference density")
```
```suggestion
    ref_density = xs.variable(description="reference saturation water vapor density")
```

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
-    winddir = xs.variable(description="wind direction")
-    tau_c = xs.variable(description="conversion time")
-    tau_f = xs.variable(description="fallout time")
+    lat = xs.variable(description="latitude")
```
I would slightly prefer `latitude`.

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -9,21 +10,19 @@
+    precip_base = xs.variable(description="background precipitation")
+    wind_speed = xs.variable(description="windspeed")
+    wind_dir = xs.variable(description="wind direction")
+    conv_time = xs.variable(description="conversion time")
+    fall_time = xs.variable(description="fallout time")
```
```suggestion
    conv_time = xs.variable(description="cloud water to hydrometeor conversion time")
    fall_time = xs.variable(description="hydrometeor fallout time")
```

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -58,13 +60,16 @@ def run_step(self):
-    runoff = xs.foreign(OrographicPrecipitation, 'precip_rate', attrs={"units": "mm/h"})
-    discharge = xs.on_demand(dims=('y','x'), description='discharge from orographic precipitation', attrs={"units": "mm/h"})
+class OrographicDrainageDischarge(FlowAccumulator):
+    """Flowaccumulation including orographic precipitation
```
Let me suggest an alternative description:

```
Accumulate orographic precipitation from upstream to downstream.

For use in the context of landscape evolution modeling, ``flowacc`` values
are converted from mm^3 h^-1 to m^3 yr^-1. For convenience, the ``discharge``
on demand variable still returns the values in mm^3 h^-1.  
```

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -57,4 +60,29 @@ def run_step(self):
+    def run_step(self):
+        super().run_step()
+
+        # scale mm/h to m/yr
+        self.flowacc *= 8.76
```
We actually want to scale from mm^3/h to m^3/yr, so the conversion factor should be 8.76e-6 instead.

That said, it makes me wonder if absolute (scaled) values of discharge are the right values to be used directly as input for computing stream power channel erosion:

- [Fastscapelib-fortran's docs](https://fastscape-lem.github.io/fastscapelib-fortran/#_fastscape_set_precip) suggests using relative values for runoff (i.e., `runoff = OrographicPrecipitation.precip_rate / OrographicPrecipitation.precip_base`), since the spatially averaged precipitation magnitude is already included implicitly in the parameters of the stream power law.

- I'm wondering whether using discharge per unit with would make more sense, as in the classic stream power law we already use drainage area expressed in m^2.

Any thoughts on this @jeanbraun ?

benbovy reviewed (commented):
This looks good, I like the variable names!

I left a few comments.

We should now define some default values according to a "standard atmosphere". The last paragraph of the appendix of Smith & Barstad paper may be helpful.

We should also address what kind of flow accumulated values is given to the channel erosion process (see my comment below).

rlange2 committed:
use consistent variable names (plus some variable description)

rlange2 committed:
remove attr import

Originally, I thought attrs would be needed to use custom metadata, such
as units.

rlange2 committed:
update variable names in process

rlange2 committed:
move calculation of cw out of dictionary

rlange2 committed:
update variable description and metadata

rlange2 committed:
Process OrographicDrainageDischarge has been updated

Docstring includes a proper description of the process. Metadata have
been added to variable discharge. Conversion of flowacc and discharge
accounts for the whole drainage area.

rlange2 committed:
supply default values

rlange2 commented:
The process `OrographicPrecipitation` is now equipped with default values as can be found in the mentioned appendix section of the original paper. Next, I will work on an extended docstring that contains how these values have been derived and supply additional working ranges for the majority of input variables.

rlange2 committed:
update docstring

rlange2 committed:
fix typo

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -1,31 +1,90 @@
     """Computes orographic precipitation following Smith & Barstad (2004)
 
-    Returns precipitation rate in m/yr
+    Attributes
```
The Attributes section will be added by Xarray-simlab, with `autodoc` ;-). No need to add it here.

It still makes sense to add value ranges in the docstrings, though. You could create a `Notes` section for that (numpydoc has no strict convention for that section).

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -1,31 +1,90 @@
+                             attrs={"units": "m/s"})
+    wind_dir = xs.variable(description="wind direction (azimuth)",
+                           attrs={"units": "degrees"})
+    conv_time = xs.variable(description="conversion time",
```
Any default value for `conv_time`, `fall_time` and `cw`?

It would be nice if the parameters with a required value could be limited to `latitude`, `precip_base`, `wind_dir` and `wind_speed`.

benbovy reviewed (commented):
Looks good. I only left two minor comments.

benbovy commented on orographic_precipitation/fastscape_ext.py:
```diff
@@ -34,18 +93,23 @@ class OrographicPrecipitation:
+            "cw" : self.cw}
 
     def initialize(self):
         self._params = self._get_params()
```
Remove this line, we don't need parameters in the initialize stage.

benbovy replied:
In `run_step`, you could simply do

```python
self.precip_rate = compute_orographic_precip(self.elevation,
                                             self.dx,
                                             self.dy,
                                             **self._get_params())
```

benbovy commented:
Don't forget to update `README` with the new parameter names set here for `compute_orographic_precip`.

rlange2 committed:
update README with new var names

rlange2 committed:
include PR's in testing

rlange2 committed:
update notes in docstring

rlange2 committed:
add test for process OrographicPrecipitation

rlange2 committed:
update dependencies

rlange2 committed:
update dependencies

benbovy commented:
This looks good to me, I think you can go ahead and merge this PR. We can add more tests in another one. Thanks Raphael!
